gadget/validate.go

```diff
@@ -230,31 +240,73 @@ func ensureRolesConsistency(roles map[string]*roleInstance, expectedSeed bool) e
 	return nil
 }
 
+func ensureRolesConsistencyOnModalClassic(roles map[string]*roleInstance) error {
+	if roles[SystemBoot] == nil || roles[SystemData] == nil {
+		return fmt.Errorf("system-boot and system-data roles are needed on classic")
+	}
+
+	// Make sure labels are as expected - save is optional
+	roleLabelToCheck := []roleLabel{
+		{SystemBoot, ubuntuBootLabel}, {SystemData, ubuntuDataLabel}}
+	roleLabelOptional := []roleLabel{
+		{SystemSeed, ubuntuSeedLabel}, {SystemSave, ubuntuSaveLabel}}
+	for _, rlLb := range roleLabelOptional {
+		if roles[rlLb.role] != nil {
+			roleLabelToCheck = append(roleLabelToCheck, rlLb)
+		}
+	}
+	if err := checkImplicitLabels(roleLabelToCheck, roles); err != nil {
+		return err
+	}
+
+	// Check that boot/seed/save are in the same volume
+	bootVolName := roles[SystemBoot].volName
+	for _, rlLb := range roleLabelOptional {
+		if roles[rlLb.role] != nil {
+			if roles[rlLb.role].volName != bootVolName {
+				return fmt.Errorf("system-boot and %s are expected to share the same volume", rlLb.role)
+			}
+		}
+	}
+	return nil
+}
+
 func ensureSystemSaveRuleConsistency(roles map[string]*roleInstance) error {
 	if roles[SystemData] == nil || roles[SystemSeed] == nil {
 		// previous checks should stop reaching here
 		return fmt.Errorf("internal error: system-save requires system-seed and system-data structures")
 	}
-	if err := checkImplicitLabel(SystemSave, roles[SystemSave].s, ubuntuSaveLabel); err != nil {
+	if err := checkImplicitLabels([]roleLabel{roleLabelSave}, roles); err != nil {
 		return err
 	}
 	return nil
 }
 
 func checkSeedDataImplicitLabels(roles map[string]*roleInstance) error {
-	if err := checkImplicitLabel(SystemData, roles[SystemData].s, ubuntuDataLabel); err != nil {
-		return err
-	}
-	if err := checkImplicitLabel(SystemSeed, roles[SystemSeed].s, ubuntuSeedLabel); err != nil {
+	if err := checkImplicitLabels([]roleLabel{roleLabelData, roleLabelSeed}, roles); err != nil {
 		return err
 	}
 	return nil
 }
 
-func checkImplicitLabel(role string, vs *VolumeStructure, implicitLabel string) error {
-	if vs.Label != "" && vs.Label != implicitLabel {
-		return fmt.Errorf("%s structure must have an implicit label or %q, not %q", role, implicitLabel, vs.Label)
+// roleLabel contains a partition role and the default expected label.
+type roleLabel struct {
+	role  string
+	label string
+}
+
+var (
+	roleLabelSeed = roleLabel{role: SystemSeed, label: ubuntuSeedLabel}
+	roleLabelSave = roleLabel{role: SystemSave, label: ubuntuSaveLabel}
+	roleLabelData = roleLabel{role: SystemData, label: ubuntuDataLabel}
+)
 
+func checkImplicitLabels(roleLabels []roleLabel, roles map[string]*roleInstance) error {
+	for _, rlLb := range roleLabels {
+		volStruct := roles[rlLb.role].s
```

Comment:
I wish that `s` had a longer name, but this is preexisting code, so nevermind :-)

Reply:
this should return an internall error now if volStruct is nil. The older function signature didn't have this problem as much as it was much clearer that it would be called with a non-nil volume structure

Reply:
Check added, also function changed to make it variadic that seemed more natural.